Feat: move rust code to a dedicated subfolder + add metadata for ALL Valkey commands

mindreframer opened the pull request:
Just some preparation for future changes.

mindreframer committed:
Feat: group all rust crates into a single folder

Problem:
- currently the folder structure is rather flat, and all types of folders are in the root folder. this makes it a bit harder to quickly skim and see what the purpose for each folder is

Solution:
- move rust folders to "crates"

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Feat: copy all command descriptions from Valkey to Sabledb

Problem:
- since we aim for compatibility with Valkey/REDIS, it would be useful to have declarative metadata for all supported commands

Solution:
- just copy JSON files from https://github.com/valkey-io/valkey/tree/unstable/src/commands
- this allows for more structured comparison for the compatibility

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Chore: further cleanup the top-folder

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Feat: Bun.js script to generate commands overview in a single markdown document

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Feat: generate sabledb-support.jsonl file

Problem:
- we need to track, which command is already supported in SableDB

Solution:
- keep a separate small jsonl file, where we store the info about support for each command

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Feat: generated compatibility overview and store in "docs" folder

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Feat: update docs/COMPATIBILITY.md document

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Feat: update commands support from commander.rs

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

sabledb-io commented:
Notice that the `test_loading_cert` test is failing, it is probably due to the restructuring, since it is using a relative path to load the test certificate.

mindreframer committed:
Fix: update paths for SSL certificate tests

Problem:
- after folder restructuring our SSL tests are failing
- the paths to load them from was not adjusted

Solution:
- adjust the path

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer commented:
I also get this error: 

---- commands::string_commands::test::test_string_commands::psetex stdout ----
Error in command: get psetex_key1

thread 'commands::string_commands::test::test_string_commands::psetex' panicked at crates/libsabledb/src/commands/string_commands.rs:1451:25:
assertion `left == right` failed
  left: "$-1\r\n"
 right: "$5\r\nvalue\r\n"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

but it was already on main branch before my changes.

sabledb-io commented:
I would guess it is flaky.

The test case there is:

```rust
        (vec!["psetex", "psetex_key1", "42", "value"], "+OK\r\n"),
        (vec!["get", "psetex_key1"], "$5\r\nvalue\r\n"),
```

We first set `psetex_key1` with `42ms` expiry and we then call `get` - if 42ms passed between the two - then the item was already expired and we will get an empty string `$-1\r\n` (it can happen since many tests are running in parallel so it might slow the system)

sabledb-io commented:
Anyways, I enabled all the tests (including cluster integration tests) so lets see if they pass on GitHub (they are passing on multiple platform locally)

sabledb-io commented:
You can try and run this test locally:

```rust
cargo test commands::string_commands::test::test_string_commands::psetex
```

and i suspect that it will pass this time for you

sabledb-io commented on support/compatibility_check/bun.lock:
I am not TS expert, but are we suppose to add `.lock` files to `git`?

mindreframer replied:
yes, that is the purpose of this file, go guarantee reproducible builds. 
otherwise it's YOLO in the land of JS / TS.

sabledb-io commented on support/compatibility_check/codebase-commands.json:
This file contains some hard coded paths from your computer - lets clean this.

sabledb-io commented on support/compatibility_check/extract-from-commander.ts:
Can we run this script using some kind of custom build target? (we can add this as a step to the `build.rs`)

mindreframer replied:
why not? it's up to you. it's a single bash command. you can do anything. bun installs on Github CI in couple seconds, then you can do whatever. it's really very flexible and i use it for high level code generation instead of clunky bash scripts.

sabledb-io replied:
I would generate the file: [COMPATIBILITY.md](https://github.com/sabledb-io/sabledb/pull/35/files#diff-0333285b158fbef700dbdf89f364f7edbada5846a67ab9d5888e056c8ea98a1d) during build time (locally) and link it from the main `README.md` file. It provides a valuable information about the project status.

mindreframer committed:
Chore: replace absolute paths with relative in codebase-commands.json

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Chore: remove codebase-commands.json (this is not actively used anymore)

Run: `bun run extract-from-commander.ts` instead
Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer committed:
Chore: make file names more consistent

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer commented:
> and i suspect that it will pass this time for you

indeed, it passes when run in isolation. but yeah, if you are fine with some level of flakiness, then we can leave it as it is.

mindreframer commented:
could you activate automatic CI workflows, otherwise the feedback loop for contributions is needlessly slowed down.

sabledb-io commented:
> could you activate automatic CI workflows, otherwise the feedback loop for contributions is needlessly slowed down.

The default was: "Require approval for first-time contributors" I changed it into "Require approval for first-time contributors who are new to GitHub" (it is the most permissive that GH allows)

mindreframer committed:
Fix: now proper fix for SSL cert paths

Signed-off-by: Roman Heinrich <roman.heinrich@gmail.com>

mindreframer commented:
@sabledb-io is there any way to chat in a more realtime way with you? Like discord or telegram? 

I would like to discuss some ideas, and i feel using github issues for this would be quite time-consuming.

sabledb-io commented:
> is there any way to chat in a more realtime way with you? Like discord or telegram?

I think we can setup a gitter.im through GH. Let me check (I have this for my other open-source projects...)

sabledb-io commented:
Try this room: https://app.gitter.im/#/room/#sabledb:gitter.im
